### ISSUE TITLE
Unwanted Warning in lr_scheduler.step()

### ISSUE THREAD

**Issue by Avishka-Perera** [pytorch/pytorch#117540]
### 🐛 Describe the bug

I need a fine learning rate variation over the minibatches in my training loop. Hence I do the following:

```python
import numpy as np
import torch
from torch.nn import Linear
from torch.optim import SGD
from torch.optim.lr_scheduler import CosineAnnealingLR
import matplotlib.pyplot as plt

epoch_count = 10
mini_batch_count = 20

model = Linear(10, 20)
optm = SGD(model.parameters(), 1e-3)
schd = CosineAnnealingLR(optm, T_max=epoch_count)

dl = torch.Tensor(mini_batch_count, 32, 10)

lrs = []
for e in range(epoch_count):
    for batch_id, batch in enumerate(dl):
        model(batch)
        optm.step()
        lrs.append(schd.get_last_lr()[0])
        schd.step(e + (1 + batch_id) / mini_batch_count)

plt.plot(lrs)
plt.title("Smooth variation of the LR")
plt.xticks(np.arange(0, epoch_count*mini_batch_count+1, mini_batch_count), np.arange(0, epoch_count+1))
plt.show()
```

But this gives the warning "The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose."

The only way to silence this warning is to call the `step()` method without the epoch parameter, but it does not satisfy my requirement.

```
import torch
from torch.nn import Linear
from torch.optim import SGD
from torch.optim.lr_scheduler import CosineAnnealingLR
import matplotlib.pyplot as plt

epoch_count = 10
mini_batch_count = 20

model = Linear(10, 20)
optm = SGD(model.parameters(), 1e-3)
schd = CosineAnnealingLR(optm, T_max=epoch_count)

dl = torch.Tensor(mini_batch_count, 32, 10)

lrs = []
for e in range(epoch_count):
    for batch_id, batch in enumerate(dl):
        model(batch)
        optm.step()
        lrs.append(schd.get_last_lr()[0])
    schd.step()

plt.plot(lrs)
plt.title("Coarse variation of the LR")
plt.xticks(
    np.arange(0, epoch_count * mini_batch_count + 1, mini_batch_count),
    np.arange(0, epoch_count + 1),
)
plt.show()
```

Thanks 

### Versions

PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.11.7 (main, Dec 15 2023, 18:12:31) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-169-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Quadro RTX 6000
GPU 1: Quadro RTX 6000

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             40
On-line CPU(s) list:                0-39
Thread(s) per core:                 2
Core(s) per socket:                 10
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz
Stepping:                           7
CPU MHz:                            1644.417
CPU max MHz:                        3200.0000
CPU min MHz:                        1000.0000
BogoMIPS:                           4400.00
Virtualization:                     VT-x
L1d cache:                          640 KiB
L1i cache:                          640 KiB
L2 cache:                           20 MiB
L3 cache:                           27.5 MiB
NUMA node0 CPU(s):                  0-9,20-29
NUMA node1 CPU(s):                  10-19,30-39
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: Split huge pages
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.1
[pip3] numpy==1.26.2
[pip3] pytorch-lightning==2.1.3
[pip3] pytorch-msssim==1.0.0
[pip3] segmentation-models-pytorch==0.3.3
[pip3] torch==2.1.2
[pip3] torchmetrics==1.2.1
[pip3] torchvision==0.16.2
[pip3] triton==2.1.0
[conda] efficientnet-pytorch      0.7.1                    pypi_0    pypi
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] pytorch-lightning         2.1.3                    pypi_0    pypi
[conda] pytorch-msssim            1.0.0                    pypi_0    pypi
[conda] segmentation-models-pytorch 0.3.3                    pypi_0    pypi
[conda] torch                     2.1.2                    pypi_0    pypi
[conda] torchmetrics              1.2.1                    pypi_0    pypi
[conda] torchvision               0.16.2                   pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

**Comment by janeyx99** [comment]
Hello, thank you for reporting your use case. I have a naïve question (sorry I haven't looked much into LRSchedulers): what would be the difference between your current use case with `schd.step(e + (1 + batch_id) / mini_batch_count)` and the following?

```
lrs = []
for e in range(epoch_count):
    for batch_id, batch in enumerate(dl):
        model(batch)
        optm.step()
        lrs.append(schd.get_last_lr()[0])
        schd.step()
```

I'd like to understand your use case a bit better. From looking at recent issues, the epoch deprecation handling has been confusing due to not having a dedicated expert on LRSchedulers, so I am hoping to learn more about your use case to determine next steps.

**Comment by Avishka-Perera** [comment]
Hi Jane,

This the expected variation of the LR throughout the 10 epochs (generated from the code in my initial post that produces the warning)
![fb5ffc31-b8c5-4d7c-9e5d-c663639b8c00](https://github.com/pytorch/pytorch/assets/74392492/f93042d9-ab66-4266-a996-8105bf1c9fd6)

This is the variation throughout the 10 epochs if I use the proposed method
![acd74adf-7985-4605-864a-67936c3dfd37](https://github.com/pytorch/pytorch/assets/74392492/8a65617d-b733-4151-a7f7-f956df5e880c)

Given below is the exact code that I used to generate it.
```python
import numpy as np
import torch
from torch.nn import Linear
from torch.optim import SGD
from torch.optim.lr_scheduler import CosineAnnealingLR
import matplotlib.pyplot as plt

epoch_count = 10
mini_batch_count = 20

model = Linear(10, 20)
optm = SGD(model.parameters(), 1e-3)
schd = CosineAnnealingLR(optm, T_max=epoch_count)

dl = torch.Tensor(mini_batch_count, 32, 10)

lrs = []
for e in range(epoch_count):
    for batch_id, batch in enumerate(dl):
        model(batch)
        optm.step()
        lrs.append(schd.get_last_lr()[0])
        schd.step()

plt.plot(lrs)
plt.title("Unwanted variations")
plt.xticks(
    np.arange(0, epoch_count * mini_batch_count + 1, mini_batch_count),
    np.arange(0, epoch_count + 1),
)
plt.show()
```

**Comment by janeyx99** [comment]
Sorry it took me this long to get here! Feel free to continue using the epoch kwarg for now as we currently do not have review bandwidth/active maintainers to properly deprecate it.
